apps/web/src/services/dashboards/stores/dashboard-main-page-store.ts

```diff
@@ -1,17 +1,138 @@
+import { store } from '@/store';
+
+import { useDashboardStore } from '@/store/dashboard/dashboard-store';
+
+import type { DashboardTreeDataType } from '@/services/dashboards/types/dashboard-folder-type';
+
+
+type DashboardModel = PublicDashboardModel | PrivateDashboardModel;
+type FolderModel = PublicFolderModel | PrivateFolderModel;
+const _getDashboardTreeData = (folderList: FolderModel[], dashboardList: DashboardModel[]): TreeNode<DashboardTreeDataType>[] => {
+    const nodes: Record<string, TreeNode<DashboardTreeDataType>> = {};
+    folderList.forEach((d) => {
+        nodes[d.folder_id] = {
+            id: d.folder_id,
+            depth: 0,
+            data: {
+                name: d.name,
+                type: 'FOLDER',
+                id: d.folder_id,
+            },
+            children: [],
+        };
+    });
+    dashboardList.forEach((d) => {
+        nodes[d.dashboard_id] = {
+            id: d.dashboard_id,
+            depth: 0,
+            data: {
+                name: d.name,
+                id: d.dashboard_id,
+                type: 'DASHBOARD',
+                folderId: d.folder_id,
+                shared: d?.shared,
+                scope: d?.scope,
+                userId: d?.user_id,
+                createdBy: d.tags?.created_by,
+            },
+        };
+    });
+
+    const rootNodes: TreeNode<DashboardTreeDataType>[] = [];
+    const setDepth = (node, depth) => {
+        node.depth = depth;
+        if (!node.children) return;
+        node.children.forEach((child) => {
+            setDepth(child, depth + 1);
+        });
+    };
+    Object.values(nodes).forEach((node) => {
+        const folderId = node.data?.folderId;
+        if (!folderId) {
+            rootNodes.push(node);
+            setDepth(node, 0);
+        } else {
+            const parentNode = nodes[folderId];
+            if (parentNode) {
+                parentNode.children = parentNode.children || [];
+                parentNode.children.push(node);
+                setDepth(node, parentNode.depth + 1);
+            }
+        }
+    });
+
+    return rootNodes;
+};
+const _getSelectedTreeData = (dashboardTreeData: TreeNode<DashboardTreeDataType>[], selectedIdMap: Record<string, boolean>): TreeNode<DashboardTreeDataType>[] => {
+    const _selectedIdList = Object.keys(selectedIdMap).filter((key) => selectedIdMap[key]);
+    const _selectedNodeList: TreeNode<DashboardTreeDataType>[] = [];
+    _selectedIdList.forEach((id) => {
+        const _node = dashboardTreeData.find((node) => node.data.id === id);
+        if (!_node) return;
+        if (_node?.data.type === 'FOLDER') {
+            // get folder id only if all children are selected
+            if (_node.children?.every((child) => selectedIdMap[child.id])) {
+                _selectedNodeList.push(_node);
+            } else {
+                const _childrenDashboards = _node.children?.filter((child) => selectedIdMap[child.id]);
```

Comment:
suggestion: I think it's pretty inefficient..🥲 
what about to add flag like `removeFolder`, and do not use `Array.prototype.every()` method in the if statement? 
it loops twice for else cases, and I think most cases will be the else case.